File: Offer 14-I.go
package leetcode

import "math"

func cuttingRope(n int) int {
	// n为的最大值为dp[j] * dp(n - j)
	dp := make([]int, n+1)
	if n <= 2 {
		return 1
	}
	max := func(a, b  int) int {
		if a >= b {
			return a
		} else {
			return b
		}
 	}
	dp[2] = 1
	for i := 3; i <= n; i++ {
		for j := 1; j <= i; j++ {
			dp[i] = max(dp[i], max(j * (i - j), (i - j) * dp[j]))
		}
	}
	return dp[n]
}

File: Offer 14-I.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int cuttingRope(int n) {
        if (n <= 2) return 1;
        vector<int> dp(n + 1);
        dp[2] = 1;        
        for (int i = 3; i <=n; ++i) {
            for (int j = 1; j < i; ++j) {
                dp[i] = max(dp[i], max(j * (i - j), j * dp[i-j]));
            }
        }
        return dp.back();
    }
};